src/gaborkernels.jl

```diff
@@ -0,0 +1,323 @@
+
+```math
+g(x, y) = \\exp(-\\frac{x'^2 + \\gamma^2y'^2}{2\\sigma^2})\\exp(i(2\\pi\\frac{x'}{\\lambda} + \\psi))
+```
+
+where ``x' = x\\cos\\theta + y\\sin\\theta`` and ``y' = -x\\sin\\theta + y\\cos\\theta``.
+
+# References
+
+- [1] [Wikipedia: Gabor filter](https://en.wikipedia.org/wiki/Gabor_filter)
+- [2] [Wikipedia: Gabor transformation](https://en.wikipedia.org/wiki/Gabor_transform)
+- [3] [Wikipedia: Gabor atom](https://en.wikipedia.org/wiki/Gabor_atom)
+- [4] [Gabor filter for image processing and computer vision](http://matlabserver.cs.rug.nl/edgedetectionweb/web/edgedetection_params.html)
+"""
+struct Gabor{T<:Complex, TP<:Real, R<:AbstractUnitRange} <: AbstractMatrix{T}
+    ax::Tuple{R,R}
+    λ::TP
+    θ::TP
+    ψ::TP
+    σ::TP
+    γ::TP
+
+    # cache values
+    sc::Tuple{TP, TP} # sincos(θ)
+    λ_scaled::TP # 2π/λ
+    function Gabor{T,TP,R}(ax::Tuple{R,R}, λ::TP, θ::TP, ψ::TP, σ::TP, γ::TP) where {T,TP,R}
+        λ > 0 || throw(ArgumentError("`λ` should be positive: $λ"))
+        new{T,TP,R}(ax, λ, θ, ψ, σ, γ, sincos(θ), 2π/λ)
+    end
+end
+function Gabor(size_or_axes::Tuple; wavelength, orientation, kwargs...)
+    Gabor(size_or_axes, wavelength, orientation; kwargs...)
+end
+function Gabor(
+        size_or_axes::Tuple,
+        wavelength::Real,
+        orientation::Real;
+        bandwidth::Real=1.0f0,
+        phase_offset::Real=0.0f0,
+        aspect_ratio::Real=0.5f0,
+    )
+    bandwidth > 0 || throw(ArgumentError("`bandwidth` should be positive: $bandwidth"))
+    aspect_ratio > 0 || throw(ArgumentError("`aspect_ratio` should be positive: $aspect_ratio"))
+    wavelength >= 2 || @warn "`wavelength` should be equal to or greater than 2" wavelength
+    # we still follow the math symbols in the implementation
+    λ, γ, ψ = wavelength, aspect_ratio, phase_offset
+    # for orientation: Julia follow column-major order, thus here we compensate the orientation by π/2
+    θ = convert(float(typeof(orientation)), mod2pi(orientation+π/2))
+    # follows reference [4]
+    σ = convert(float(typeof(λ)), (λ/π)*sqrt(0.5log(2)) * (2^bandwidth+1)/(2^bandwidth-1))
+
+    params = float.(promote(λ, θ, ψ, σ, γ))
+    T = typeof(params[1])
+
+    ax = _to_axes(size_or_axes)
+    all(map(length, ax) .> 0) || throw(ArgumentError("Kernel size should be positive: $size_or_axes"))
+    if any(r->5λ > length(r), ax)
+        expected_size = @. 5λ * length(ax)
+        @warn "for wavelength `λ=$λ` the expected kernel size is expected to be larger than $expected_size."
+    end
+
+    Gabor{Complex{T}, T, typeof(first(ax))}(ax, params...)
+end
+
+@inline Base.size(kern::Gabor) = map(length, kern.ax)
+@inline Base.axes(kern::Gabor) = kern.ax
+@inline function Base.getindex(kern::Gabor, x::Int, y::Int)
+    ψ, σ, γ = kern.ψ, kern.σ, kern.γ
+    s, c = kern.sc # sincos(θ)
+    λ_scaled = kern.λ_scaled # 2π/λ
+
+    xr = x*c + y*s
+    yr = -x*s + y*c
+    yr = γ * yr
+    return exp(-(xr^2 + yr^2)/(2σ^2)) * cis(xr*λ_scaled + ψ)
```

Comment:
What happens if we index this with out-of-bounds `x` and `y`? If lack of `BoundsError` is deliberate, it should be documented.